src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs

```diff
@@ -0,0 +1,514 @@
+        /// <returns>True if a LibraryType is Project or ExternalProject.</returns>
+        private static bool IsProject(LibraryType type)
+        {
+            return (type == LibraryType.ExternalProject || type == LibraryType.Project);
+        }
+
+        /// <summary>
+        /// A simple node class to hold the outgoing dependency edge during the graph walk.
+        /// </summary>
+        private class DependencyNode : IEquatable<DependencyNode>
+        {
+            // ID of the Node 
+            public string Id { get; }
+
+            // bool to indicate if the node is a project node
+            // if false then the node is a package
+            public bool IsProject { get; }
+
+            // WarningPropertiesCollection of the path taken to the Node
+            public WarningPropertiesCollection WarningPropertiesCollection { get; }
+
+            public DependencyNode(string id, bool isProject, WarningPropertiesCollection warningPropertiesCollection)
+            {
+                Id = id ?? throw new ArgumentNullException(nameof(id));
+                WarningPropertiesCollection = warningPropertiesCollection ?? throw new ArgumentNullException(nameof(warningPropertiesCollection));
+                IsProject = isProject;
+            }
+
+            public DependencyNode(string id, bool isProject)
+            {
+                Id = id ?? throw new ArgumentNullException(nameof(id));
+                IsProject = isProject;
+            }
+
+            public override int GetHashCode()
+            {
+                var hashCode = new HashCodeCombiner();
+
+                hashCode.AddStringIgnoreCase(Id);
+                hashCode.AddObject(IsProject);
+
+                return hashCode.CombinedHash;
+            }
+
+            public override bool Equals(object obj)
+            {
+                return Equals(obj as DependencyNode);
+            }
+
+            public bool Equals(DependencyNode other)
+            {
+                if (other == null)
+                {
+                    return false;
+                }
+
+                if (ReferenceEquals(this, other))
+                {
+                    return true;
+                }
+
+                return string.Equals(Id, other.Id, StringComparison.OrdinalIgnoreCase) &&
+                    IsProject == other.IsProject &&
+                    WarningPropertiesCollection.Equals(other.WarningPropertiesCollection);
+            }
+
+            public override string ToString()
+            {
+                return $"{(IsProject ? "Project" : "Package")}/{Id}";
+            }
+        }
+
+        /// <summary>
+        /// A simple node class to hold the outgoing dependency edge for a quick look up
+        /// </summary>
+        private class LookUpNode
+        {
+            // List of dependencies for this node
+            public IEnumerable<LibraryDependency> Dependencies { get; set; }
```

Comment:
This is `IEnumerable` because it is read from `RemoteResolveResult.Dependencies` which is also an `IEnumerable`.